[IGNORE] review datasource documentation (#1490)
perses/perses: docs/api/datasource.md

# Datasource

> ## Different Datasource scope

Reviewer: ```suggestion
## Choose a scope
```

> We have three different ways to define a datasource, depending on the scope we want to give

Reviewer: ```suggestion
There are 3 different scopes in which you can define a datasource, depending at which point you want it to be shared.
- for common use cases, use higher scopes in order to reuse the same datasource on multiple dashboards
- for more specific needs, use lower scopes to restrict the datasource availability to a specific set of (or even a single) dashboard(s).
```

# Datasource

## Choose a scope

> There are three different scopes in which you can define a datasource, depending on which point you want it to be shared.

Reviewer: ```suggestion
There are three different scopes in which you can define a datasource, depending on how much you want it to be shared.
```

### Project level

In case you would like to share a datasource across different dashboards in the **same** project, you will need to
create a Datasource.

```yaml
  kind: "Datasource"
  metadata:
    name: <string>
    project: <string>
  spec: <datasource_spec>
```

### Global level

When we talk about scope and user permission in a REST API, the easiest way is to associate one permission per endpoint.

> If we want to provide a datasource shared by all projects, then it makes sense to create a different object that is

Reviewer: ```suggestion
If we want to provide a datasource shared by all projects, then it makes sense to have a different object that is
```

living outside a project.

> That’s why we will have a new resource called `GlobalDatasource`

Reviewer: ```suggestion
That’s why we have another resource called `GlobalDatasource`
```

  spec: <plugin_spec>
```

We are supporting only prometheus as a datasource for the moment.
Please look at the [documentation](./plugin/prometheus.md#datasource) to know the spec for the Prometheus datasource.

### Selecting / Referencing a Datasource

In the panels, you will be able to select a datasource. Like proposed in the first draft, the selector will be like
this:

```typescript
interface DatasourceSelector {
  kind: string;
  // name is the unique name of the datasource to use
  // If name is omitted, that effectively means "use the default datasource for this kind".
  name?: string;
}
```

Priority is `"Local datasource in the dashboard" > "Project datasource" > "Global datasource"`.

So if by any chance you have a local datasource that is named exactly like a Project datasource, or a Global datasource,
we will consider that the user intentionally wanted to override the upper datasource. We will use the one with the
smallest scope.

### How to use the Perses' proxy

As described before, you can provide a proxy configuration that will be used by the Perses server to redirect any
queries to the datasource.

> It means in case of the Prometheus datasource, if the field `directUrl` is not set, then the FE needs to use the Perses

Reviewer: ```suggestion
It means in the case of the Prometheus datasource, if the field `directUrl` is not set then the FE needs to use the Perses
```

```yaml
  kind: "GlobalDatasource"
  metadata:
    name: <string>
  spec: <datasource_spec>
```

## Datasource specification

```yaml
  [ display: <display_spec> ]

  # If true, then it's the default datasource for the type defined in the plugin.
  [ default: <boolean> | default = false ]

  # The definition of the plugin datasource
  plugin: <plugin>
```

### Plugin definition

```yaml
  # The type of the datasource. For example, `PrometheusDatasource`
  kind: <string>

  # The actual definition of the datasource. It will depend on the type defined in the previous field `kind`
  spec: <plugin_spec>
```

We are supporting only prometheus as a datasource for the moment.
Please look at the [documentation](./plugin/prometheus.md#datasource) to know the spec for the Prometheus datasource.

### Selecting / Referencing a Datasource

In the panels, you will be able to select a datasource. Like proposed in the first draft, the selector will be like
this:

```typescript
interface DatasourceSelector {
  kind: string;
  // name is the unique name of the datasource to use
  // If name is omitted, that effectively means "use the default datasource for this kind".
  name?: string;
}
```

> Priority is `"Local datasource in the dashboard" > "Project datasource" > "Global datasource"`.

Reviewer: ```suggestion
The priority order is `"Local datasource in the dashboard" > "Project datasource" > "Global datasource"`.
```
